### PR TITLE
PR: Add position to text completion

### LINKED ISSUE #8723: autocomplete fails when pressing tab too quickly
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
autocomplete fails when pressing tab too quickly

### What steps reproduce the problem?

1. in the editor, type 
```
import numpy as np
np.new
```
2. `<tab><enter>`

### What is the expected output? What do you see instead?
if `<tab><enter>` pressed quickly:
```
np.new
newaxis
```
if more slowly:
`np.newaxis`

## Versions

* Spyder version: Spyder 4.0.0.dev0 (last master)
Python 3.7.0 64-bit | Qt 5.9.3 | PyQt5 5.9.2 | Darwin 18.2.0

### LINKED ISSUE #8727: autocomplete fails to select an option in the list
### Issue Report Checklist

* [ ] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
When the autocomplete options are displayed, a letter won't select an option

### What steps reproduce the problem?
```
import numpy as np
np.ne
```
`<tab>`
A list appears, with `np.negative` selected, and `np.newaxis`, which is what I want.
`w`
Nothing happens.

### What is the expected output? What do you see instead?
I expect `np.newaxis` to be selected

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: 
* Python version: 
* Qt version: 
* PyQt version: 
* Operating System name/version: 

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)
* [ ] Included a screenshot (if affecting the UI)

While an autocomplete is being processed, the cursor can move. Auto completing on another line is clearly not what is expected. The cursor is therefore moved back if this happens. 
Also, the selection of an option in the list was broken.

### Issue(s) Resolved

Fixes #8727 
Fixes #8723 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
Impact27

### CODE REVIEW

File: spyder/plugins/editor/widgets/base.py [hunk 1]
```diff
@@ -183,16 +184,30 @@ def keyPressEvent(self, event):
...
+            self.sendKeyToTextEdit(event)
             self.update_current()
         elif modifier:
-            self.textedit.keyPressEvent(event)
+            self.sendKeyToTextEdit(event)
         else:
             self.hide()
             QListWidget.keyPressEvent(self, event)
 
+    def sendKeyToTextEdit(self, event):
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/base.py]: Please change this method's name to `send_key_to_textedit`.

File: spyder/plugins/editor/widgets/base.py [hunk 2]
```diff
@@ -183,16 +184,30 @@ def keyPressEvent(self, event):
...
+        """Send key to text edit and updates the position of
+        the text completion"""
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/base.py]: Please rewrite this to

```
"""
Send key to text edit and updates the position of
the text completion.
"""
```

**Comment by impact27** [comment]
I changed it a bit: the sendKeyToTextEdit didn't really make sense because the cursor should not be before the position where you started autocompleting. Therefore I changed it so that it closes if you press backspace.

**Comment by impact27** [comment]
Anyway that makes more sense, if I write:
`np.newb`
and I want `np.newaxis`, if I press backspace I will want a new request anyway. (So new tab)

**Comment by impact27** [comment]
One issue that is still there is:
`import numpy as np\nnp.ne<tab><backspace>\n`
and then `<enter>` when the list appears.
This will place the `negative` on the next line (because the position changed as a letter was removed).
I am not sure it is a big issue as that would only apply if somebody messes with the characters before the completion position In the time between the `<tab>` and the reply from LSP.

**Comment by impact27** [comment]
I would need some help with the tests. 
what I want to test is:
 1. Complete `math.a <tab>` ... `s <enter>` to `math.asin`
This will check if asin can be selected in the list
 2. Complete `math.a <tab><enter>` ... `<enter>` to `math.acos`
This will check if when enter is pressed too rapidly, the correction is still correct
 3. Complete` math.a <tab> s` ...`<enter>` to `math.asin`
This will check if when s is pressed too rapidly, asin is still selected

**Comment by impact27** [comment]
I fixed the tests

**Comment by ccordoba12** [comment]
Thanks! This looks really good but I'm going to ask @dalthviz to give it a final review since he has worked on code completion a lot.

**Comment by dalthviz** [comment]
Hi @impact27 thanks for working in this! While reviewing I encountered an error:

```python
Traceback (most recent call last):
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\lsp\client.py", line 230, in on_msg_received
    handler(resp['result'], req_id)
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\lsp\providers\document.py", line 117, in process_document_completion
    LSPRequestTypes.DOCUMENT_COMPLETION, {'params': response})
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\widgets\codeeditor.py", line 690, in handle_response
    handler(params)
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\widgets\codeeditor.py", line 791, in process_completion
    self.completion_widget.show_list(completion_list, position)
  File "C:\Users\Daniel\Desktop\spyder_others\spyder-impact\spyder\spyder\plugins\editor\widgets\base.py", line 64, in show_list
    if self.textedit.textCursor().position() < position:
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```

I was trying to complete `import num -> import numpy` 

The error appeared only the first time I tried (I'm unsure of how `position` could had a `None` value). After restarting Spyder it worked without issue. Could you check what is happening or maybe add a validation to prevent position being `None`? Thanks :)

**Comment by impact27** [comment]
I have added a validation. Maybe the function is called twice?

**Comment by impact27** [comment]
<tab><tab> does the trick

**Comment by ccordoba12** [comment]
@impact27, please merge with master to get the fixes to our tests I pushed yesterday.

**Comment by ccordoba12** [comment]
Thanks @impact27 for your work and @dalthviz for the review!

**Pull request merged**